_posts/2021-06-07-Synchronizing-physical-clocks-among-different-Subnero-modems.md

```diff
@@ -0,0 +1,87 @@
+
+There are two major problems to be addressed to synchronize clocks among different modems:
+
+1. Modems start at different times so there may be a *clock offset* as shown in figure below:
+
+![Clock offset between two modems]({{site.baseurl}}/images/clock-offset.png){: .center-image }
+
+2. Modem clocks tick at different rates causing a *clock drift* as shown in figure below:
+
+![Clock drift between two modems]({{site.baseurl}}/images/clock-drift.png){: .center-image }
+
+
+# Compensating for clock offset
+
+The `Ranging` agent in UnetStack running on the modem generates the clock offset information based on the message exchanges when range measurement is performed on the modem that initiates ranging. The clock offset between the two modems is populated in the ranging agent's parameter `ranging[host('B')].offset`. Where `host('B')` is the node address of modem B. More details on this can be found ![here](https://unetstack.net/handbook/unet-handbook_ranging_and_synchronization.html). 
+
+Once, the clock offset is known, it can be used to adjust the time locally to synchronize the physical time (`phy.time`). However, if a low-drift clock is not used, there is a possibility that the clocks tick at different rates, and clocks drift faster. Let us take a look at how we can compensate for the clock drift next.
+
+# Compensating for clock drift
+
+Since the modems might tick at different rates, one modem's clock may be faster than the other. An example is shown in the figure above where Modem A's clock is faster than Modem B, i.e., a slope of 1.2 on Modem A's clock. To compensate for this, the user of the modem can utilize a `phy.clockCalib` parameter. This parameter is a multiplier on the physical clock. So its default value is 1. If you set it to 1.2 the clock will run at 1.2x the rate of the actual hardware clock.
+
+Therefore, to compensate Modem A's clock to run slower so that it matches Modem B's clock, we can set `phy.clockCalib = 1/1.2 = 0.83`.
+
+
+The users of the modem should make measurements to figure out the nomimal values that they should set for `phy.clockCalib`. 
+
+# An example of calibrating clock
+
+We will consider two modems A and B for this example. To synchronize modem A's clock with the modem B's clock the following procedure may be followed:
+
+1.  Make a measurement of system time and physical clock time on modem A.
+
+To measure system time, open modem's web shell and type the following:
+
+```groovy
+> phy.rtc.getTime()
+
+```
+
+To measured physical clock time, open modem's web shell and type the following:
+
+```groovy
+> phy.time
+
+```
+
+`phy.rtc.getTime()` returns the number of milliseconds since January 1, 1970, 00:00:00 GMT represented by this `Date` object (`phy.rtc`). The `phy.time` returns time in microseconds. Make a record of both these measurements in the same unit.
```

Comment:
How? Perhaps we should show a simple script that someone can run, and then just get the clock calibration number?

Reply:
I tried to add a code snippet to show a way to do this now. Please check if you think that makes sense.

Reply:
I still don't see how someone can follow it step-by-step. By the time they do one step and the next, time has elapsed. You need a single script that does this.

Reply:
New conversation on this below.

Reply:
I have provided a script below to run on say modem A and modem B to figure out the slope based on which calibration factor to set in `phy.clockCalib` can be computed.

Reply:
OK